src/java/org/apache/cassandra/db/commitlog/CommitLogSegmentManagerCDC.java

```diff
@@ -190,14 +189,20 @@ public CommitLogSegment.Allocation allocate(Mutation mutation, int size) throws
         return alloc;
     }
 
-    // Non-blocking mode has just recently been enabled for CDC.
-    // The segment is still marked as FORBIDDEN. It should be set to PERMITTED.
-    private void ensureSegmentPermittedIfNotBlockWrites(CommitLogSegment segment)
+    // Permit a forbidden segment under the following conditions.
+    // - Non-blocking mode has just recently been enabled for CDC.
+    // - The CDC total space has droppped below the limit (e.g. CDC consumer cleans up).
+    private void permitSegmentMaybe(CommitLogSegment segment)
     {
-        if (!DatabaseDescriptor.getCDCBlockWrites() && segment.getCDCState() == CDCState.FORBIDDEN)
+        if (segment.getCDCState() != CDCState.FORBIDDEN)
+            return;
+
+        if (!DatabaseDescriptor.getCDCBlockWrites()
+            || cdcSizeTracker.sizeInProgress.get() < DatabaseDescriptor.getCDCTotalSpace())
```

Comment:
The `CDCSizeTracker` sums up the size of all commit logs, including CDC and non-CDC files, since it creates hard links for all newly created files. (A side question is.. maybe we want to created the link unless the state is `FORBIDDEN`?). So the size is often an over-estimate when the workload contains both CDC and non-CDC traffic. 
Back to the scenario in `allocate`... if the state was previously `FORBIDDEN`, it means the total size has exceeded the limit at the time of creation. Now, the total size drops below the limit. The `CDCSizeTracker` must have run at least once. In this case, the size of the file (`allocatingFrom`) is already counted.

Reply:
No. If the the state is FORBIDDEN, we do not create a link. The link is created only when the state is NOT FORBIDDEN (i.e. PERMITTED/CONTAINS). 
The current implementation is that a hardlink is created for all new segments, regardless of the state. Since each FORBIDDEN segment has a change to revert back to PERMITTED, this handling is easy but at the cost of slightly over-counting (the size of one file). 
I think, and as you mentioned, it likely not a big problem.

Reply:
I'm confused. Here's what I'm seeing in the code:
```
    public CommitLogSegment createSegment()
    {
        CommitLogSegment segment = CommitLogSegment.createSegment(commitLog, this);

        // Hard link file in cdc folder for realtime tracking
        FileUtils.createHardLink(segment.logFile, segment.getCDCFile());

        cdcSizeTracker.processNewSegment(segment);
        return segment;
    }
```
I think the code would need to read something like this to prevent hard-linking files with cdc disabled:
```
    public CommitLogSegment createSegment()
    {
        CommitLogSegment segment = CommitLogSegment.createSegment(commitLog, this);

        if (segment.getCDCState() != CDCState.FORBIDDEN)
        {
            // Hard link file in cdc folder for realtime tracking
            FileUtils.createHardLink(segment.logFile, segment.getCDCFile());
        }

        cdcSizeTracker.processNewSegment(segment);
        return segment;
    }
```

Further, we'd need to add logic to hard link the file on transition from FORBIDDEN to PERMITTED if that happened later in its lifecycle.

That make sense?

Reply:
It makes sense. The if statement is what I proposed in the earlier comment. The ease of handling in the current implementation is what I guessed by reading the code.